src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -236,156 +234,142 @@ public unsafe bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte
-            {
-                RejectRequestHeader(headerLine, length);
-            }
-
-            // Skip colon from value start
+            var headerName = headerLine[..nameEnd];
+            // Skip colon to get to the value start.
             var valueStart = nameEnd + 1;
-            // Ignore start whitespace
-            for (; valueStart < valueEnd; valueStart++)
+            // Ignore start whitespace. Generally there will only be one space
+            // so we will just do a char by char loop inline.
+            do
             {
-                var ch = headerLine[valueStart];
-                if (ch != ByteTab && ch != ByteSpace && ch != ByteCR)
+                // Should be a for/while loop https://github.com/dotnet/runtime/issues/9422
+                // Test in if to drop range check for following span access
+                if ((uint)valueStart >= (uint)headerLine.Length)
                 {
                     break;
                 }
-                else if (ch == ByteCR)
+                var ch = headerLine[valueStart];
+                if (ch != ByteTab && ch != ByteSpace)
                 {
-                    RejectRequestHeader(headerLine, length);
+                    // Non-whitespace char found, valueStart is now start of value.
+                    break;
                 }
-            }
 
-            // Check for CR in value
-            var valueBuffer = new Span<byte>(headerLine + valueStart, valueEnd - valueStart + 1);
-            if (valueBuffer.Contains(ByteCR))
-            {
-                RejectRequestHeader(headerLine, length);
-            }
+                valueStart++;
+            } while (true);
 
-            // Ignore end whitespace
-            var lengthChanged = false;
+            // Ignore end whitespace. Generally there will no spaces
+            // so we will just do a char by char loop from end inline.
+            var valueEnd = headerLine.Length - 1;
             for (; valueEnd >= valueStart; valueEnd--)
             {
                 var ch = headerLine[valueEnd];
                 if (ch != ByteTab && ch != ByteSpace)
                 {
+                    // Non-whitespace char found, valueEnd is now start of value.
                     break;
                 }
-
-                lengthChanged = true;
             }
 
-            if (lengthChanged)
-            {
-                // Length changed
-                valueBuffer = new Span<byte>(headerLine + valueStart, valueEnd - valueStart + 1);
-            }
+            // Range end is exclusive, so add 1 to valueEnd
+            valueEnd++;
+            var headerValue = headerLine[valueStart..valueEnd];
 
-            var nameBuffer = new Span<byte>(headerLine, nameEnd);
+            handler.OnHeader(headerName, headerValue);
+            return true;
 
-            handler.OnHeader(nameBuffer, valueBuffer);
+        Reject:
+            // Reject is a jump forward as we expect most headers to be accepted,
```

Comment:
```suggestion
            // Reject is a conditional jump forward as we expect most headers to be accepted,
```
I'd like to precise this if we comment on branch predictor's behavior and afaik (at least on Intel) unconditional jumps don't accord to the rule of "forward jumps".

Reply:
Added